libs/cardano-ledger-binary/src/Cardano/Ledger/Binary/Decoding/Decoder.hs

```diff
@@ -831,6 +835,20 @@ decodeMapNoDuplicates decodeKey decodeValue =
       !value <- decodeValue
       pure (key, value)
 
+-- | Just like `decodeMap`, but fails on duplicate keys
+decodeMapEnforceNoDuplicates ::
+  Ord k => Decoder s k -> (k -> Decoder s v) -> Decoder s (Map.Map k v)
+decodeMapEnforceNoDuplicates decodeKey decodeValueFor = do
+  (numValues, xs) <- decodeCollectionWithLen decodeMapLenOrIndef decodeInlinedPair
+  unless (numValues == (length . Set.toList . fromList . (map fst)) xs) $
+    fail "Map has duplicates"
+  pure $ Map.fromList xs
```

Comment:
This is way too expensive:
* construct a list
* construct another list with keys
* construct a set from a list
* construct a map from a list

There is a much better way to do this.

Construct a Map and each time a key value pair is decoded check weather it is already in a Map or not.
*Disclaimer* - In PR programming, expect some problems, but the idea should be clear.

```suggestion
  decodeMapLenOrIndef >>= \case
    Just len -> loop (\(x -> pure (x - 1, x  <= 0)) len Map.empty
    Nothing -> loop (\x -> (,) x <$> decodeBreakOr) () Map.empty
  where
    loop condition prevStep !acc = do
      (nextStep, shouldStop) <- condition prevStep
      if shouldStop
        then pure acc
        else do
          (k, v) <- decodeInlinedPair
          when (k `Map.member` acc) $ fail "Duplicate key detected in the Map"
          loop condition nextStep (Map.insert k v acc)
```

Reply:
awesome, thank you!